File: src/web/apps/file-manager/src/progress-stages.ts
/**
 * The pipeline's steps, and how far through a file each one is.
 *
 * The order here mirrors the `ProgressStage` enum in
 * contracts/webview-messages-v1.json, which is the source of the vocabulary —
 * `test/progress-stages.test.ts` fails if the two drift apart. The host states
 * which step a message belongs to, so nothing here reads the message text.
 *
 * Weights are this file's own, and they are presentation: they say how much of
 * a file's wait each step tends to be, which is what makes one bar advance
 * evenly instead of restarting per step. They are rough by nature; only their
 * proportions matter.
 */

export interface StageDefinition {
  /** What the pane calls this step. */
  readonly label: string;
  /** Share of a file's total wait, relative to the other stages. */
  readonly weight: number;
}

/**
 * Declared in pipeline order. A run skips steps — a document with a trusted
 * text layer never reaches `ocr` — so this is a ranking, not a sequence to
 * expect in full, and progress jumps over what did not run.
 */
export const STAGES: ReadonlyMap<string, StageDefinition> = new Map([
  ["queue",           { label: "Queued",            weight: 0 }],
  ["prepare",         { label: "Preparing",         weight: 1 }],
  ["convert",         { label: "Converting",        weight: 3 }],
  ["transfer",        { label: "Sending",           weight: 3 }],
  ["security",        { label: "Securing",          weight: 4 }],
  ["source-check",    { label: "Checking text",     weight: 5 }],
  ["source",          { label: "Using text",        weight: 1 }],
  ["geometry",        { label: "Reading text",      weight: 6 }],
  ["ocr",             { label: "Recognizing",       weight: 40 }],
  ["orientation",     { label: "Checking rotation", weight: 4 }],
  ["adaptive-ocr",    { label: "Refining",          weight: 15 }],
  ["table-recovery",  { label: "Reading tables",    weight: 8 }],
  ["table-structure", { label: "Finding tables",    weight: 6 }],
  ["fs-structure",    { label: "Reading structure", weight: 3 }],
  ["values",          { label: "Finding values",    weight: 3 }],
  ["result-transfer", { label: "Saving",            weight: 3 }],
  ["finalizing",      { label: "Finalizing",        weight: 2 }],
]);

export const STAGE_ORDER: readonly string[] = Array.from(STAGES.keys());

const TOTAL_WEIGHT = Array.from(STAGES.values()).reduce((sum, s) => sum + s.weight, 0);

/** Weight of every stage declared before this one. */
const WEIGHT_BEFORE: ReadonlyMap<string, number> = (() => {
  const before = new Map<string, number>();
  let running = 0;
  for (const [stage, definition] of STAGES) {
    before.set(stage, running);
    running += definition.weight;
  }
  return before;
})();

/** The pane's name for a stage, or null when the host names one we don't know. */
export function stageLabel(stage: string | undefined): string | null {
  if (!stage) return null;
  return STAGES.get(stage)?.label ?? null;
}

/**
 * How far through the whole file a stage sits, as 0–1.
 *
 * `current`/`total` count within the stage only, so they scale that stage's own
 * span rather than driving the bar directly. That is the difference that
 * matters: the counts arriving during OCR are pages, during transfer they are
 * chunks, and feeding either straight to the bar is what made it fill and reset
 * several times per file.
 *
 * Returns null for a stage we have no weight for, so the caller can leave the
 * bar where it was instead of guessing.
 */
export function fileProgressFraction(
  stage: string | undefined,
  current?: number,
  total?: number,
): number | null {
  if (!stage) return null;
  const definition = STAGES.get(stage);
  const before = WEIGHT_BEFORE.get(stage);
  if (!definition || before === undefined || TOTAL_WEIGHT <= 0) return null;

  let within = 0;
  if (typeof current === "number" && typeof total === "number" && total > 0) {
    within = Math.min(Math.max(current / total, 0), 1);
  }

  return Math.min(Math.max((before + definition.weight * within) / TOTAL_WEIGHT, 0), 1);
}

File: src/DocuLink.Addin/Modules/Infrastructure/ProgressStages.cs
namespace DocuLink.Addin.Modules.Infrastructure
{
    /// <summary>
    /// The steps a job moves through, mirroring the <c>ProgressStage</c> enum in
    /// contracts/python-worker-v1.json and contracts/webview-messages-v1.json.
    /// The contract is the source of the vocabulary; this is the host's copy of it,
    /// and the three must be changed together.
    ///
    /// <see cref="Order"/> is pipeline order, which is what lets the file manager
    /// place a step on a bar. An adaptive run skips steps, so it is a ranking
    /// rather than a sequence to expect in full.
    ///
    /// Queue, Prepare, Transfer and Finalizing are the host's own; the worker
    /// states the rest.
    /// </summary>
    internal static class ProgressStages
    {
        public const string Queue = "queue";
        public const string Prepare = "prepare";
        public const string Convert = "convert";
        public const string Transfer = "transfer";
        public const string Security = "security";
        public const string SourceCheck = "source-check";
        public const string Source = "source";
        public const string Geometry = "geometry";
        public const string Ocr = "ocr";
        public const string Orientation = "orientation";
        public const string AdaptiveOcr = "adaptive-ocr";
        public const string TableRecovery = "table-recovery";
        public const string TableStructure = "table-structure";
        public const string FsStructure = "fs-structure";
        public const string Values = "values";
        public const string ResultTransfer = "result-transfer";
        public const string Finalizing = "finalizing";

        public static readonly string[] Order =
        {
            Queue,
            Prepare,
            Convert,
            Transfer,
            Security,
            SourceCheck,
            Source,
            Geometry,
            Ocr,
            Orientation,
            AdaptiveOcr,
            TableRecovery,
            TableStructure,
            FsStructure,
            Values,
            ResultTransfer,
            Finalizing,
        };
    }
}
